### PR TITLE
Enable SSE support for Cloud using staged rollout (GB feature flag)

### PR DESCRIPTION
**Pull request by bryce-fitzsimons**
### Features and Changes
Cloud:
- Streaming will no longer be a paid feature, and will be available for **everybody**.
- The rollout will be controlled (by organization), dogfooded using GB feature flag rules.

### CODE REVIEW

File: packages/back-end/src/controllers/features.ts [hunk 1]
```diff
@@ -143,6 +146,20 @@ async function getPayloadParamsFromApiKey(
...
+  if (IS_CLOUD) {
+    growthbook = new GrowthBook<AppFeatures>({
+      apiHost: "https://cdn.growthbook.io",
+      clientKey:
+        process.env.NODE_ENV === "production"
+          ? "sdk-ueFMOgZ2daLa0M"
+          : "sdk-UmQ03OkUDAu7Aox",
+      enableDevMode: true,
+    });
+    await growthbook.loadFeatures({ autoRefresh: false });
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/controllers/features.ts:160]: This works locally. Do we need to use a Fetch polyfill?

I also considered pulling from mongo, but not sure if it's worth the effort given this is usually CDN cached.

File: packages/back-end/src/controllers/features.ts [hunk 2]
```diff
@@ -143,6 +146,20 @@ async function getPayloadParamsFromApiKey(
 }
 
 export async function getFeaturesPublic(req: Request, res: Response) {
+  // Cloud only: Invoke GB for SSE rollout rules
+  let growthbook: GrowthBook | null = null;
+  if (IS_CLOUD) {
+    growthbook = new GrowthBook<AppFeatures>({
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/controllers/features.ts:152]: Admittedly a bit inelegant. We could move this to app outside of the request/response lifecycle, but not entirely sure its worth it. Feedback welcome.

File: packages/back-end/src/controllers/features.ts (hunk 1 above)
Comment (bryce-fitzsimons) [reply packages/back-end/src/controllers/features.ts:160]: any advice for avoiding the infinite CDN->features loop? Do you suggest manually building the payload from mongo when its our own?

Comment (bryce-fitzsimons) [reply packages/back-end/src/controllers/features.ts:160]: any advice for avoiding the infinite CDN->features loop? Do you suggest manually building the payload from mongo when its our own?

File: packages/back-end/src/controllers/features.ts [hunk 3]
```diff
@@ -156,13 +173,32 @@ export async function getFeaturesPublic(req: Request, res: Response) {
...
+        : "unknown";
+
+      growthbook.setAttributes({
+        company: currentOrg?.name || "",
+        organizationId: hashedOrganizationId,
+        cloud: IS_CLOUD,
+        accountPlan: accountPlan,
+        freeSeats: currentOrg?.freeSeats || 3,
+        discountCode: currentOrg?.discountCode || "",
+      });
```
[the remark this answers is not in the document]
Comment (bryce-fitzsimons) [reply packages/back-end/src/controllers/features.ts:182]: I'm not sure it's identical, just similar. For the public features endpoint, only a subset of attributes were readily available, and some (such as `url`) didn't make sense. Settled on a standard middleware for most routes, and manually setting the attributes that play nicely with the features endpoint (in terms of availability, speed, CDN cache-ability).

File: packages/back-end/src/services/GBCloudSDKService/middleware.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      company: req?.organization?.name || "",
+      organizationId: req?.organization?.id
+        ? sha256(req.organization.id, GROWTHBOOK_SECURE_ATTRIBUTE_SALT)
+        : "",
+      url: req.get("Referrer") || "",
+      cloud: IS_CLOUD,
+      accountPlan: req?.organization
+        ? getAccountPlan(req?.organization) || "unknown"
+        : "unknown",
+      hasLicenseKey: !!getLicense(),
```
Comment (bryce-fitzsimons) [inline packages/back-end/src/services/GBCloudSDKService/middleware.ts:71]: this one might not be correct, at least for cloud users. Is there another way to get the license?

File: packages/back-end/src/services/GBCloudSDKService/middleware.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  configureCache({ backgroundSync: false });
+
+  res.on("close", () => {
+    gb.destroy();
+  });
+
+  try {
+    await promiseTimeout(
+      // Do not subscribe individual SDK instances to streaming updates
+      gb.loadFeatures({ autoRefresh: false }),
```
Comment (jdorn) [inline packages/back-end/src/services/GBCloudSDKService/middleware.ts:44]: There's a built-in `timeout` prop.  Is there a reason we're not using that?

File: packages/back-end/src/services/GBCloudSDKService/middleware.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    apiHost: "https://cdn.growthbook.io",
+    clientKey:
+      process.env.NODE_ENV === "production"
+        ? "sdk-ueFMOgZ2daLa0M"
+        : "sdk-UmQ03OkUDAu7Aox",
+  });
+  req.app.locals.growthbook = gb;
+
+  // Disable SSE
+  configureCache({ backgroundSync: false });
```
Comment (jdorn) [inline packages/back-end/src/services/GBCloudSDKService/middleware.ts:35]: This needs to happen before we instantiate a `GrowthBook` instance, otherwise it will have no effect

File: packages/front-end/components/Features/SDKConnections/SDKConnectionsList.tsx [hunk 7]
```diff
@@ -80,7 +84,8 @@ export default function SDKConnectionsList() {
             {connections.map((connection) => {
               const hasProxy =
                 !isCloud() && connection.proxy.enabled && connection.proxy.host;
-              const hasCloudProxyForSSE = isCloud() && connection.sseEnabled;
+              const hasCloudProxyForSSE =
+                isCloud() && growthbook?.isOn("proxy-cloud-sse");
```
Comment (jdorn) [inline packages/front-end/components/Features/SDKConnections/SDKConnectionsList.tsx:88]: We're not using this feature flag anymore, right?

File: packages/back-end/src/services/GBCloudSDKService/middleware.ts (hunk 5 above)
Comment (bryce-fitzsimons) [reply packages/back-end/src/services/GBCloudSDKService/middleware.ts]: I moved all this to https://github.com/growthbook/growthbook/pull/1505/files and addressed your feedback

File: packages/front-end/components/Features/SDKConnections/SDKConnectionsList.tsx (hunk 7 above)
Comment (bryce-fitzsimons) [reply packages/front-end/components/Features/SDKConnections/SDKConnectionsList.tsx]: we're using this temporarily until 100% rollout, just so we can show accurate info in the UI

File: packages/back-end/src/models/SdkConnectionModel.ts [hunk 8]
```diff
@@ -69,13 +68,13 @@ const SDKConnectionModel = mongoose.model<SDKConnectionInterface>(
...
   return {
     ...proxy,
-    enabled: PROXY_ENABLED,
-    host: PROXY_HOST_INTERNAL || PROXY_HOST_PUBLIC,
-    hostExternal: PROXY_HOST_PUBLIC || PROXY_HOST_INTERNAL,
+    enabled: proxy.enabled !== undefined ? proxy.enabled : PROXY_ENABLED,
+    host: proxy.host ? proxy.host : PROXY_HOST_INTERNAL || PROXY_HOST_PUBLIC,
+    hostExternal: proxy.hostExternal
+      ? proxy.hostExternal
+      : PROXY_HOST_PUBLIC || PROXY_HOST_INTERNAL,
```
Comment (jdorn) [inline packages/back-end/src/models/SdkConnectionModel.ts:77]: What's the purpose of this change?  I think this might break self-hosted users who are using a proxy with environment variables.

File: packages/back-end/src/models/SdkConnectionModel.ts [hunk 9]
```diff
@@ -149,15 +147,14 @@ export async function createSDKConnection(params: CreateSDKConnectionParams) {
     key: generateSDKConnectionKey(),
     proxy: addEnvProxySettings({
       enabled: !!proxyEnabled,
-      host: proxyHost || "",
-      hostExternal: proxyHost || "",
+      host: proxyHost || PROXY_HOST_INTERNAL || PROXY_HOST_PUBLIC,
+      hostExternal: proxyHost || PROXY_HOST_PUBLIC || PROXY_HOST_INTERNAL,
```
Comment (jdorn) [inline packages/back-end/src/models/SdkConnectionModel.ts:151]: Is this needed?  We're calling `addEnvProxySettings` here, so this should be added already there

File: packages/back-end/src/models/SdkConnectionModel.ts (hunk 8 above)
Comment (bryce-fitzsimons) [reply packages/back-end/src/models/SdkConnectionModel.ts:77]: I believe the original function breaks our "local proxy for cloud users" functionality: They need to be able to set `enabled` and `host` from the SDK Connection UI, but then this method overwrites these fields with env var defaults. And even if they aren't actually set in our cloud environment, we'd still overwrite with `""`. We need a way to only set these fields if nothing is being provided by `...proxy`.

I don't fully understand the point of `hostExternal` so I'm not confident that I got this part correct.

**Review by jdorn — APPROVED** [review]
